src/include/processor/operator/persistent/delete_executor.h

```diff
@@ -9,61 +9,59 @@ namespace processor {
     virtual ~NodeDeleteExecutor() = default;
 
-    void init(ResultSet* resultSet, ExecutionContext* context);
+    virtual void init(ResultSet* resultSet, ExecutionContext* context);
 
-    virtual void delete_() = 0;
+    virtual void delete_(ExecutionContext* context) = 0;
 
     virtual std::unique_ptr<NodeDeleteExecutor> copy() const = 0;
 
 protected:
     DataPos nodeIDPos;
-    DataPos primaryKeyPos;
-
     common::ValueVector* nodeIDVector;
-    common::ValueVector* primaryKeyVector;
 };
 
 class SingleLabelNodeDeleteExecutor : public NodeDeleteExecutor {
 public:
-    SingleLabelNodeDeleteExecutor(
-        storage::NodeTable* table, const DataPos& nodeIDPos, const DataPos& primaryKeyPos)
-        : NodeDeleteExecutor(nodeIDPos, primaryKeyPos), table{table} {}
+    SingleLabelNodeDeleteExecutor(storage::NodeTable* table, const DataPos& nodeIDPos)
+        : NodeDeleteExecutor(nodeIDPos), table{table} {}
     SingleLabelNodeDeleteExecutor(const SingleLabelNodeDeleteExecutor& other)
-        : NodeDeleteExecutor(other.nodeIDPos, other.primaryKeyPos), table{other.table} {}
+        : NodeDeleteExecutor(other.nodeIDPos), table{other.table} {}
 
-    void delete_() final;
+    void init(ResultSet* resultSet, ExecutionContext* context) final;
+    void delete_(ExecutionContext* context) final;
 
     inline std::unique_ptr<NodeDeleteExecutor> copy() const final {
         return std::make_unique<SingleLabelNodeDeleteExecutor>(*this);
     }
 
 private:
     storage::NodeTable* table;
+    std::unique_ptr<storage::NodeTable::DeleteState> deleteState;
 };
 
 class MultiLabelNodeDeleteExecutor : public NodeDeleteExecutor {
 public:
     MultiLabelNodeDeleteExecutor(
         std::unordered_map<common::table_id_t, storage::NodeTable*> tableIDToTableMap,
-        const DataPos& nodeIDPos, const DataPos& primaryKeyPos)
-        : NodeDeleteExecutor(nodeIDPos, primaryKeyPos), tableIDToTableMap{
-                                                            std::move(tableIDToTableMap)} {}
+        const DataPos& nodeIDPos)
+        : NodeDeleteExecutor(nodeIDPos), tableIDToTableMap{std::move(tableIDToTableMap)} {}
     MultiLabelNodeDeleteExecutor(const MultiLabelNodeDeleteExecutor& other)
-        : NodeDeleteExecutor(other.nodeIDPos, other.primaryKeyPos), tableIDToTableMap{
-                                                                        other.tableIDToTableMap} {}
+        : NodeDeleteExecutor(other.nodeIDPos), tableIDToTableMap{other.tableIDToTableMap} {}
 
-    void delete_() final;
+    void init(ResultSet* resultSet, ExecutionContext* context) final;
+    void delete_(ExecutionContext* context) final;
 
     inline std::unique_ptr<NodeDeleteExecutor> copy() const final {
         return std::make_unique<MultiLabelNodeDeleteExecutor>(*this);
     }
 
 private:
     std::unordered_map<common::table_id_t, storage::NodeTable*> tableIDToTableMap;
+    std::unordered_map<common::table_id_t, std::unique_ptr<storage::NodeTable::DeleteState>>
```

Comment:
We should probably wrap NodeTable and DeleteState into a pair/struct so that we don't need to do 2 lookup at runtime

Reply:
Yeah, struct or pair has better semantic here, though I wouldn't worry about the extra map lookup. I did try to wrap them into a struct, but awkwardly ended up with that now two fields (table, deleteState) of the struct are separately initialized in the constructor and `init()` call, which is also not a good design. I don't have a better idea in my mind for now, so will keep this as it is and we will revisit as we figure out how to simplify the multi-label and single-label design here.